test/system/flows/join_track_test.rb

```diff
@@ -0,0 +1,20 @@
+require "application_system_test_case"
+
+module Flows
+  class JoinTrackTestTest < ApplicationSystemTestCase
+    test "joins track when authenticated" do
+      user = create :user
+      track = create :track, slug: "ruby", title: "Ruby"
+
+      # TODO: Add when adding devise
+      # sign_in(user)
+
+      visit tracks_url
+      click_link "Ruby"
+
+      click_on "Join Track"
+
+      assert_equal [track], user.tracks
```

Comment:
Can we instead assert on text instead of values in the database? I view system tests from a user perspective, and a user shouldn't have direct access to the database.

Reply:
I think we want two different things here. And I think probably we want a test for each.

- I want something that uses the app as a user does but where we verify that the correct things are happening under the hood. Without this test, there's nothing in our test-suite that actually checks that things work as expected. These are effectively integration tests, but I want them to use the Chrome for the testing, not me manually specifying the routes to traverse (which I find prone to bugs).
- You want something that tests the flow of the app specifically from how a user is perceiving it. That's really valuable too.

My issue is when we combine the two, making view-level changes, break the "functional" test that I want, which I find really frustrating. And that's even more acute at this early stage, as the actual views are effectively nonsense, but I still want to have proper functional tests that don't then break when we implement the views.

---

So maybe the key here is to have the tests in two different subdirectories where it's clear what the purpose of each is?

Reply:
Moved to https://github.com/exercism/v3-website/issues/54